sql/parser/builtins.go

```diff
@@ -62,103 +62,92 @@ type builtin struct {
 	// TODO(XisiHuang): support encoding, i.e., length(str, encoding).
-	"length": stringBuiltin(func(s string) (Datum, error) {
+	"length": {stringBuiltin(func(s string) (Datum, error) {
 		return DInt(len(s)), nil
-	}),
+	})},
 
-	"lower": stringBuiltin(func(s string) (Datum, error) {
+	"lower": {stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToLower(s)), nil
-	}),
+	})},
 
-	"upper": stringBuiltin(func(s string) (Datum, error) {
+	"upper": {stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToUpper(s)), nil
-	}),
+	})},
 
 	// TODO(XisiHuang): support the substring(str FROM x FOR y) syntax.
 	"substring": {
-		nArgs: -1,
-		fn: func(args DTuple) (Datum, error) {
-			argsNum := len(args)
-			if argsNum != 2 && argsNum != 3 {
-				return DNull, fmt.Errorf("incorrect number of arguments: %d vs %s",
-					argsNum, "2 or 3")
-			}
-
-			dstr, ok := args[0].(DString)
-			if !ok {
-				return DNull, argTypeError(args[0], dstr.Type())
-			}
-			str := string(dstr)
-
-			start, ok := args[1].(DInt)
-			if !ok {
-				return DNull, argTypeError(args[1], start.Type())
-			}
-			s := int(start)
-
-			var e int
-			if argsNum == 2 {
-				e = len(str) + 1
-			} else {
-				count, ok := args[2].(DInt)
-				if !ok {
-					return DNull, argTypeError(args[2], count.Type())
+		builtin{
+			types: []string{"string", "int"},
+			fn: func(args DTuple) (Datum, error) {
+				str := string(args[0].(DString))
+				// SQL strings are 1-indexed.
+				start := int(args[1].(DInt)) - 1
+
+				if start < 0 {
+					start = 0
+				} else if start > len(str) {
+					start = len(str)
+				}
+
+				return DString(str[start:]), nil
+			},
+		},
+		builtin{
+			types: []string{"string", "int", "int"},
+			fn: func(args DTuple) (Datum, error) {
+				str := string(args[0].(DString))
+				// SQL strings are 1-indexed.
+				start := int(args[1].(DInt)) - 1
+				length := int(args[2].(DInt))
+
+				if length < 0 {
+					return DNull, fmt.Errorf("negative substring length %d not allowed", length)
 				}
-				c := int(count)
-				if c < 0 {
-					return DNull, fmt.Errorf("negative substring length not allowed: %d", c)
+
+				end := start + length
+				if end < 0 {
+					end = 0
+				} else if end > len(str) {
+					end = len(str)
 				}
-				e = s + c
-			}
-
-			if e <= 1 || s > len(str) {
-				return DString(""), nil
-			}
-
-			if s < 1 {
-				s = 1
-			}
-			if e > len(str)+1 {
-				e = len(str) + 1
-			}
-			return DString(str[s-1 : e-1]), nil
+
+				if start < 0 {
+					start = 0
+				} else if start > len(str) {
+					start = len(str)
+				}
+
+				return DString(str[start:end]), nil
+			},
 		},
 	},
 
 	// concat concatenate the text representations of all the arguments.
 	// NULL arguments are ignored.
 	"concat": {
-		nArgs: -1,
-		fn: func(args DTuple) (Datum, error) {
-			var buffer bytes.Buffer
-			for _, d := range args {
-				ds, err := datumToRawString(d)
-				if err != nil {
-					return DNull, err
+		builtin{
```

Comment:
yeah, so those functions would be registered with an empty slice, rather than a nil slice.